### PR TITLE
Init commit of substrait-cpp

### PR DESCRIPTION
**Pull request by chaojun-zhang**
Init commit of substrait-cpp 

parse substrait extension yaml to cpp Type.h, Function.h and Extension.h

### CODE REVIEW

File: src/Extension.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+} // namespace
+
+std::shared_ptr<Extension> Extension::load() {
+  static const auto registry = loadDefault();
+  return registry;
+}
+
+std::shared_ptr<Extension> Extension::loadDefault() {
+  static const std::vector<std::string> extensionFiles = {
```
Comment (haiweizh) [inline src/Extension.cpp:158]: These yaml files is not migrated into substait-cpp.

File: src/tests/FunctionLookupTest.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (haiweizh) [inline src/tests/FunctionLookupTest.cpp]: Should change copyright?

File: src/tests/FunctionLookupTest.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+TEST_F(SubstraitFunctionLookupTest, string_function) {
+  testScalarFunctionLookup("like", {kString(), kString()}, "like:opt_str_str");
+  testScalarFunctionLookup(
+      "like", {Type::decode("varchar<L1>"), Type::decode("varchar<L2>")},
+      "like:opt_vchar<L1>_vchar<L2>");
+  testScalarFunctionLookup("substr", {kString(), kI32(), kI32()},
+                           "substring:str_i64_i64");
+}
```
Comment (haiweizh) [inline src/tests/FunctionLookupTest.cpp]: Should add line break.

File: src/FunctionLookup.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#pragma once
+
+#include "Extension.h"
+#include "FunctionMapping.h"
+
+namespace io::substrait {
+
+class FunctionLookup {
+public:
+  FunctionLookup(const ExtensionPtr &extension,
```
Comment (haiweizh) [inline src/FunctionLookup.h:25]: Why remove lookupFunction(const SubstraitSignaturePtr& functionSignature) and change to this function?
Cider use lookupFunction(const SubstraitSignaturePtr& functionSignature) to look up function.

Comment (haiweizh) [reply src/FunctionLookup.h:24]: Also such as SubstraitScalarFunctionLookup and SubstraitAggregateFunctionLookup.

File: src/tests/FunctionLookupTest.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        std::make_shared<FunctionLookup>(extension_, mappings_);
+    aggregateFunctionLookup_ =
+        std::make_shared<FunctionLookup>(extension_, mappings_);
+  }
+
+  void testScalarFunctionLookup(const std::string &name,
+                                const std::vector<TypePtr> &arguments,
+                                const std::string &outputSignature) {
+    const auto &functionVariant =
+        scalarFunctionLookup_->lookupScalarFunction(name, arguments);
```
Comment (haiweizh) [inline src/tests/FunctionLookupTest.cpp]: Cider not only need comparison of input type, but also need comparison of output type when looking up function.

File: src/Type.cpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+    return std::make_shared<FixedBinaryType>(length);
+  } else if (TypeTraits<TypeKind::kStruct>::typeString ==
+             baseType) {
+    return std::make_shared<StructType>(nestedTypes);
+  } else {
+    throw std::runtime_error("Unsupported substrait type: " + rawType);
+  }
+}
+
+#define SCALAR_TYPE_MAPPING(typeKind)                                \
```
Comment (haiweizh) [inline src/Type.cpp]: Code style.

File: third_party/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+add_subdirectory(googletest)
+
+set(YAML_CPP_BUILD_TESTS OFF CACHE BOOL "Enable testing")
+include_directories(yaml-cpp/include)
+add_subdirectory(yaml-cpp)
```
Comment (vibhatha) [inline third_party/CMakeLists.txt]: nit: maybe add a newline?

File: src/Extension.cpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+  return load(yamlExtensionFiles);
+}
+
+std::shared_ptr<Extension>
+Extension::load(const std::vector<std::string> &extensionFiles) {
+  auto registry = std::make_shared<Extension>();
+  for (const auto &extensionUri : extensionFiles) {
+    const auto &node = YAML::LoadFile(extensionUri);
+
+    auto &scalarFunctions = node["scalar_functions"];
```
Comment (vibhatha) [inline src/Extension.cpp]: should this be `const`?

File: src/Extension.cpp [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    const FunctionVariantPtr &functionVariant) {
+  const auto &functionVariants =
+      windowFunctionVariantMap_.find(functionVariant->name);
+  if (functionVariants != windowFunctionVariantMap_.end()) {
+    auto &variants = functionVariants->second;
+    variants.emplace_back(functionVariant);
+  } else {
+    std::vector<FunctionVariantPtr> variants;
+    variants.emplace_back(functionVariant);
+    windowFunctionVariantMap_.insert({functionVariant->name, variants});
```
Comment (vibhatha) [inline src/Extension.cpp]: should we move?
```suggestion
    windowFunctionVariantMap_.insert({functionVariant->name, std::move(variants)});
```

File: src/FunctionMapping.h [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    return aggregateFunctionMap;
+  }
+
+  virtual const FunctionMap windowMapping() const {
+    static const FunctionMap windowFunctionMap{};
+    return windowFunctionMap;
+  }
+};
+
+using FunctionMappingPtr = std::shared_ptr<const FunctionMapping >;
```
Comment (vibhatha) [inline src/FunctionMapping.h]: nit: maybe
```suggestion
using FunctionMappingPtr = std::shared_ptr<const FunctionMapping>;
```

File: src/Extension.cpp (hunk 1 above)
Comment (chaojun-zhang) [reply src/Extension.cpp:158]: it's load from substrait submodule

File: src/tests/FunctionLookupTest.cpp (hunk 2 above)
Comment (chaojun-zhang) [reply src/tests/FunctionLookupTest.cpp]: fixed

Comment (chaojun-zhang) [reply src/tests/FunctionLookupTest.cpp]: fixed

File: src/tests/FunctionLookupTest.cpp (hunk 3 above)
Comment (chaojun-zhang) [reply src/tests/FunctionLookupTest.cpp]: fixed

File: src/Type.cpp (hunk 6 above)
Comment (chaojun-zhang) [reply src/Type.cpp]: fixed

File: third_party/CMakeLists.txt (hunk 7 above)
Comment (chaojun-zhang) [reply third_party/CMakeLists.txt]: fixed

File: src/Extension.cpp (hunk 8 above)
Comment (chaojun-zhang) [reply src/Extension.cpp]: fixed

Comment (vibhatha) [reply src/Extension.cpp]: @chaojun-zhang do you think similar code segments in the following sections would need similar adjustments.

File: src/FunctionLookup.h (hunk 4 above)
Comment (chaojun-zhang) [reply src/FunctionLookup.h:25]: reverted

File: src/tests/FunctionLookupTest.cpp (hunk 5 above)
Comment (chaojun-zhang) [reply src/tests/FunctionLookupTest.cpp]: reverted

File: src/tests/FunctionLookupTest.cpp [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+                                kFp32(),
+                            },
+                            kFp32()},
+                           "divide:opt_opt_opt_fp32_fp32");
+
+  testAggregateFunctionLookup(
+      {"avg", {Type::decode("struct<fp64,i64>")}, kFp32()}, "avg:opt_fp32");
+}
+
+TEST_F(SubstraitFunctionLookupTest, avg) {}
```
Comment (haiweizh) [inline src/tests/FunctionLookupTest.cpp:105]: I think L102's case should been moved to this empty test?

**Review by haiweizh — APPROVED** [review]

**Pull request merged**